conda_lock/models/pip_repository.py

```diff
@@ -0,0 +1,42 @@
+from hashlib import sha256
+from urllib.parse import urlparse, urlunparse
+
+from conda_lock.models.package_source import PackageSource
+
+
+class PipRepository(PackageSource):
+    @property
+    def base_url(self) -> str:
+        """The base URL of the pip repository, without a URL path."""
+        full_url = urlparse(self.url)
+        return full_url.scheme + "://" + full_url.netloc
+
+    @property
+    def stripped_base_url(self) -> str:
+        """The base URL of the pip repository, without any basic auth."""
+        base_url = urlparse(self.base_url)
+        return urlunparse(
+            base_url._replace(netloc=base_url.netloc.split("@", 1)[-1])
+        )
+
+    @property
+    def name(self) -> str:
+        """Poetry solver requires a name for each repository.
+
+        We use this to match solver results back to their relevant
+        repository.
+        """
+        sha = sha256()
+        sha.update(self.url.encode("utf-8"))
+        return sha.hexdigest()
+
+    def normalize_solver_url(self, solver_url: str) -> str:
+        """Normalize the URL returned by Poetry's solver.
+
+        Poetry doesn't return URLs with URL-based basic Auth, because it gets converted to
+        header-based Basic Auth. Because of this, we have to add the auth back in here.
+        """
+        if not solver_url.startswith(self.stripped_base_url):
+            # The resolved package URL is at a different host to the repository
+            return solver_url
```

Comment:
This is an interesting edge case.

Basically this is where the repository is at `http://private-pypi.org`, but that index directs the solver to some other host, e.g. `http://packages.private-pypi.org` or `http://somewhere.completely.different`.

Currently in this case the lockfile will just use the URL from the solver as-is, without any auth.

An alternative would be to just add the auth to the URL anyway? WDYT?